File: apps/windows/TeslaSync.App/shared-surfaces/ContextMenu.Model.cs
using System.Collections.Generic;

namespace TeslaSync.App.SharedSurfaces;

/// <summary>
/// A single contextual action the menu can present — the native port of the web <c>ContextMenuItem</c>
/// interface (web/src/components/ui/ContextMenu.tsx L55-L71). Pure, immutable data passed by the caller
/// through the <see cref="IContextMenuController"/> seam: a stable <see cref="Id"/> (the web React key), the
/// inline <see cref="Label"/>, an optional leading <see cref="IconGlyph"/> (the native analogue of the web
/// <c>icon</c> ReactNode — a Segoe Fluent glyph rather than an arbitrary element so the model stays
/// Microsoft.UI-free and headlessly testable), the <see cref="OnSelected"/> action invoked on
/// click / Enter / Space (web <c>onClick</c>; the menu auto-closes first), and the two presentational flags
/// <see cref="IsDisabled"/> (rendered but non-interactive) and <see cref="IsDestructive"/> (tinted with the
/// danger token, e.g. Delete / Archive). <see cref="Shortcut"/> is the optional right-aligned hint
/// (web <c>shortcut</c>, e.g. <c>Ctrl+D</c>). The caller owns the labels / shortcut copy and is responsible
/// for localizing them — the menu chrome itself only localizes its single accessible name.
/// </summary>
public sealed class ContextMenuItem
{
    /// <summary>Creates an immutable menu item from its caller-supplied data.</summary>
    /// <param name="id">Stable identifier (web React key); must be non-empty.</param>
    /// <param name="label">Inline display label (web <c>label</c>); may be empty but not null.</param>
    /// <param name="onSelected">Action run on activation (web <c>onClick</c>); null is an inert item.</param>
    /// <param name="iconGlyph">Optional leading Segoe Fluent glyph (web <c>icon</c>).</param>
    /// <param name="isDisabled">When true the item is shown but non-interactive (web <c>disabled</c>).</param>
    /// <param name="isDestructive">When true the item is tinted with the danger token (web <c>destructive</c>).</param>
    /// <param name="shortcut">Optional right-aligned shortcut hint (web <c>shortcut</c>).</param>
    public ContextMenuItem(
        string id,
        string label,
        Action? onSelected = null,
        string? iconGlyph = null,
        bool isDisabled = false,
        bool isDestructive = false,
        string? shortcut = null)
    {
        ArgumentException.ThrowIfNullOrEmpty(id);
        ArgumentNullException.ThrowIfNull(label);

        Id = id;
        Label = label;
        OnSelected = onSelected;
        IconGlyph = iconGlyph;
        IsDisabled = isDisabled;
        IsDestructive = isDestructive;
        Shortcut = shortcut;
    }

    /// <summary>Stable identifier used as the item key (web <c>id</c>).</summary>
    public string Id { get; }

    /// <summary>Inline display label (web <c>label</c>).</summary>
    public string Label { get; }

    /// <summary>Action invoked on activation (web <c>onClick</c>); the menu closes before it runs.</summary>
    public Action? OnSelected { get; }

    /// <summary>Optional leading Segoe Fluent icon glyph (web <c>icon</c>); null renders no icon.</summary>
    public string? IconGlyph { get; }

    /// <summary>When true the item is rendered but non-interactive (web <c>disabled</c>).</summary>
    public bool IsDisabled { get; }

    /// <summary>When true the item is tinted with the danger token (web <c>destructive</c>).</summary>
    public bool IsDestructive { get; }

    /// <summary>Optional right-aligned shortcut hint (web <c>shortcut</c>); null renders none.</summary>
    public string? Shortcut { get; }
}

/// <summary>
/// The open-menu state — the native port of the web <c>MenuState</c> (web/src/components/ui/ContextMenu.tsx
/// L73-L82). Carries the <see cref="Items"/> to present, the viewport <see cref="X"/> / <see cref="Y"/> the
/// menu anchors at, the <see cref="RestoreTarget"/> that held focus when the menu opened (restored on close —
/// the web <c>restoreFocusEl</c>, kept as an opaque object so the model is Microsoft.UI-free) and a monotonic
/// <see cref="Nonce"/> so re-opening with an identical (items, x, y) still produces a distinct snapshot and
/// re-shows the menu (web <c>nonce</c>). Immutable: the controller publishes a fresh snapshot per open.
/// </summary>
public sealed class ContextMenuSnapshot
{
    /// <summary>Creates an immutable open-menu snapshot.</summary>
    /// <param name="items">The items to present (defensively copied by the controller).</param>
    /// <param name="x">Viewport x the menu anchors at (web <c>x</c>).</param>
    /// <param name="y">Viewport y the menu anchors at (web <c>y</c>).</param>
    /// <param name="nonce">Monotonic open counter (web <c>nonce</c>).</param>
    /// <param name="restoreTarget">Element that owned focus when the menu opened (web <c>restoreFocusEl</c>).</param>
    public ContextMenuSnapshot(IReadOnlyList<ContextMenuItem> items, double x, double y, long nonce, object? restoreTarget)
    {
        ArgumentNullException.ThrowIfNull(items);

        Items = items;
        X = x;
        Y = y;
        Nonce = nonce;
        RestoreTarget = restoreTarget;
    }

    /// <summary>The items the open menu presents (web <c>items</c>).</summary>
    public IReadOnlyList<ContextMenuItem> Items { get; }

    /// <summary>Viewport x the menu anchors at before any flip (web <c>x</c>).</summary>
    public double X { get; }

    /// <summary>Viewport y the menu anchors at before any flip (web <c>y</c>).</summary>
    public double Y { get; }

    /// <summary>Monotonic open counter so identical re-opens still re-render (web <c>nonce</c>).</summary>
    public long Nonce { get; }

    /// <summary>The element that owned focus when the menu opened; restored on close (web <c>restoreFocusEl</c>).</summary>
    public object? RestoreTarget { get; }
}

/// <summary>
/// Canonical metadata + i18n keys for the context-menu surface — the native mirror of the web
/// <c>ContextMenu</c> (web/src/components/ui/ContextMenu.tsx). The web primitive is a portal-rendered menu
/// host carrying exactly one localized string, its <c>role="menu"</c> accessible name
/// (<c>t('contextMenu.menuLabel', 'Context menu')</c>); every item label / shortcut is caller-supplied and
/// localized by the caller. This metadata carries the diagnostics slug the surface registers under, that one
/// render-contract i18n key/fallback (so the native surface reproduces the web copy verbatim) and the danger
/// token key the destructive item tints from. The i18n key carries the <c>translation.</c> catalog prefix the
/// WinUI resource bridge expects (the convention every shipped surface uses) and resolves against the English
/// fallback headlessly. UI-free so it is asserted without a XAML host.
/// </summary>
public static class ContextMenuRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "ContextMenu";

    /// <summary>i18n key for the menu's accessible name (web <c>contextMenu.menuLabel</c>).</summary>
    public const string MenuLabelKey = "translation.contextMenu.menuLabel";

    /// <summary>English fallback for <see cref="MenuLabelKey"/> (web second arg, verbatim).</summary>
    public const string MenuLabelFallback = "Context menu";

    /// <summary>
    /// Generated danger brush token key (apps/design/generated/windows/Tokens.xaml) the destructive item
    /// tints from — the theme-aware native analogue of the web <c>text-rose-300</c>. Resolved at the display
    /// boundary by the view; named here so the contract is asserted headlessly.
    /// </summary>
    public const string DangerBrushKey = "TsColorDangerBrush";
}

/// <summary>
/// The resolved top-left the menu is shown at after the viewport-overflow flip — a pure value so the flip is
/// unit-tested without a XAML host.
/// </summary>
/// <param name="Left">The resolved left (web adjusted <c>el.style.left</c>).</param>
/// <param name="Top">The resolved top (web adjusted <c>el.style.top</c>).</param>
public readonly record struct ContextMenuPoint(double Left, double Top);

/// <summary>
/// The viewport-overflow flip — the native port of the web measure-and-flip pass
/// (web/src/components/ui/ContextMenu.tsx L252-L270). The web component first positions the menu at
/// <c>(x, y)</c>, measures it, then flips the anchor edge in place when it would overflow: a right-edge
/// overflow re-anchors to <c>x - width</c> (clamped to the margin) and a bottom-edge overflow re-anchors to
/// <c>y - height</c>. <see cref="Resolve"/> reproduces that arithmetic exactly so the menu never opens off the
/// visible surface; the WinUI <c>MenuFlyout</c> the view shows applies the same bounds-correction natively, so
/// this resolved point is both the faithful parity behaviour and the show position the view passes. Static and
/// side-effect-free.
/// </summary>
public static class ContextMenuPlacement
{
    /// <summary>The viewport safety margin in DIPs (web <c>VIEWPORT_MARGIN = 8</c>).</summary>
    public const double ViewportMargin = 8;

    /// <summary>
    /// Resolve the menu's top-left, flipping the anchor edge when the menu would overflow the viewport — the
    /// web <c>useLayoutEffect</c> flip. With no overflow the menu stays at <paramref name="x"/> /
    /// <paramref name="y"/>; on a right-edge overflow the left re-anchors to
    /// <c>max(margin, x - menuWidth)</c>; on a bottom-edge overflow the top re-anchors to
    /// <c>max(margin, y - menuHeight)</c>. The two edges are evaluated independently, exactly as the web source
    /// adjusts left and top in separate guards.
    /// </summary>
    /// <param name="x">Requested anchor x (web <c>state.x</c>).</param>
    /// <param name="y">Requested anchor y (web <c>state.y</c>).</param>
    /// <param name="menuWidth">Measured / estimated menu width (web <c>rect.width</c>).</param>
    /// <param name="menuHeight">Measured / estimated menu height (web <c>rect.height</c>).</param>
    /// <param name="viewportWidth">Viewport width (web <c>window.innerWidth</c>).</param>
    /// <param name="viewportHeight">Viewport height (web <c>window.innerHeight</c>).</param>
    /// <param name="margin">Safety margin; defaults to <see cref="ViewportMargin"/> (web <c>VIEWPORT_MARGIN</c>).</param>
    /// <returns>The resolved, viewport-corrected top-left.</returns>
    public static ContextMenuPoint Resolve(
        double x,
        double y,
        double menuWidth,
        double menuHeight,
        double viewportWidth,
        double viewportHeight,
        double margin = ViewportMargin)
    {
        double left = x;
        double top = y;

        // web: if (left + rect.width + VIEWPORT_MARGIN > viewportW) left = Math.max(VIEWPORT_MARGIN, state.x - rect.width);
        if (left + menuWidth + margin > viewportWidth)
        {
            left = Math.Max(margin, x - menuWidth);
        }

        // web: if (top + rect.height + VIEWPORT_MARGIN > viewportH) top = Math.max(VIEWPORT_MARGIN, state.y - rect.height);
        if (top + menuHeight + margin > viewportHeight)
        {
            top = Math.Max(margin, y - menuHeight);
        }

        return new ContextMenuPoint(left, top);
    }
}

/// <summary>
/// PII-safe diagnostics for the context-menu surface (P1/S11 diagnostics contract). The menu carries
/// caller-supplied action labels that may name user data (vehicle names, saved-view titles, …), so the
/// collector records ONLY the operational <see cref="RecordViewOpened"/> signal with the surface slug — never
/// an item label, shortcut, coordinate or invoked action. Thread-safe; mirrors the shipped surfaces'
/// collectors.
/// </summary>
public sealed class ContextMenuDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">The sink the <c>view.opened</c> line is written to, or null.</param>
    public ContextMenuDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened (mounted).</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was mounted, emitting <c>view.opened slug=ContextMenu</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={ContextMenuRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/ContextMenu/ContextMenuModel.kt
// Pure, framework-free model + projection for the ContextMenu shared surface — the native analogue of every
// decision the web component makes (web/src/components/ui/ContextMenu.tsx) before Compose paints anything. No
// Compose runtime, no Android UI, no HTTP: every declaration here is exercised by the :android:testReleaseUnitTest
// gate so the composable stays a thin render layer over it.
//
// What the web source is (and therefore the COMPLETE behaviour this surface reproduces): a portal-rendered
// context-menu primitive driven by a module-level pub/sub store read through `useSyncExternalStore`. A caller
// opens it imperatively (`openContextMenu(items, x, y)`) — typically from a right-click — and the host renders a
// floating `role="menu"` at the cursor with one `role="menuitem"` per row. Rows can be disabled (rendered but
// non-interactive), destructive (tinted red), and may carry a leading icon and a right-aligned shortcut hint.
// The menu corrects viewport overflow by flipping its anchor edge (right-edge flips to x, bottom-edge flips to y)
// after the first measured layout, supports full keyboard navigation (Arrow Up/Down skip disabled rows, Home/End
// jump to first/last, Enter/Space invoke, Escape/Tab close), and refuses to open with an empty item list.
//
// Why the generic data-surface states (loading / error / stale / offline) are intentionally absent: this surface
// performs no query. Its only input is the in-process [ContextMenuStore] (the `useSyncExternalStore` analogue),
// a synchronous pub/sub that is always available and never "loads", "errors", goes "stale", or goes "offline".
// Inventing those states would model an async dependency the web spec does not have (honesty covenant: no scope
// narrowing, no silent drift). The surface's REAL, fully-reproduced states are: closed (no snapshot → the host
// renders nothing, the web `if (!snapshot) return null`), open (the menu with rows), the per-row variants
// (enabled / disabled / destructive / with-icon / with-shortcut), keyboard focus, and dismissal. The empty case
// is the web open-guard: opening with zero items is a no-op, so the menu never mounts empty and there is never a
// blank box. The keyboard-focus traversal and the viewport-flip geometry are reduced here as pure functions
// ([ContextMenuFocus], [ContextMenuPlacement]) and asserted off-device, doubling as the per-state projection.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/shared-surfaces/ContextMenu — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package intentionally
// diverges from the path — exactly as the sibling UserCell / StaggerContainer surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.contextmenu

import androidx.compose.ui.graphics.vector.ImageVector
import io.teslasync.shared.core.diagnostics.Logger
import kotlin.math.max

/** Diagnostics surface slug emitted with the `view.opened` event (P1/S11); the slug the prompt mandates. */
const val CONTEXT_MENU_SLUG: String = "ContextMenu"

/**
 * Canonical registry metadata for the ContextMenu surface. The diagnostics [SLUG] is emitted with the one-shot
 * `view.opened` event (P1/S11) and is the surface slug the prompt mandates (`ContextMenu`).
 */
object ContextMenuRegistration {
    /** Stable surface id. */
    const val ID: String = "context-menu"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = CONTEXT_MENU_SLUG
}

/**
 * One row in a context menu — the native port of the web `ContextMenuItem`
 * (`{ id, label, icon?, onClick, disabled?, destructive?, shortcut? }`). The web boolean `disabled` is modelled
 * as the inverse [enabled] (Kotlin-idiomatic, and consistent with the atomic `components/ui/ContextMenuItem`):
 * `enabled = !disabled`. A disabled row is still rendered — visibly muted and non-interactive — exactly like the
 * web `disabled` button, so the menu never silently drops rows.
 *
 * @property id stable identifier used as the list key and test tag suffix (web `item.id`).
 * @property label the text rendered inline in the row (web `item.label`).
 * @property onClick the action invoked on click / Enter / Space; the menu auto-closes first (web `item.onClick`).
 * @property enabled when false the row is shown but non-interactive (web `!item.disabled`).
 * @property destructive when true the row is tinted with the error color (web `item.destructive`, e.g. Delete).
 * @property leadingIcon optional leading glyph (web `item.icon`).
 * @property shortcut optional right-aligned shortcut hint, e.g. "⌘⇧D" (web `item.shortcut`).
 */
data class ContextMenuItem(
    val id: String,
    val label: String,
    val onClick: () -> Unit,
    val enabled: Boolean = true,
    val destructive: Boolean = false,
    val leadingIcon: ImageVector? = null,
    val shortcut: String? = null,
)

/**
 * The viewport-relative position the menu opens at, in pixels — the native port of the web `(x, y)` the
 * `onContextMenu` handler captures from `clientX` / `clientY`. Carried in window coordinates so the host's
 * popup position provider can place the menu absolutely and flip it on overflow.
 */
data class ContextMenuAnchor(
    val x: Int,
    val y: Int,
)

/**
 * The resolved top-left the menu is finally placed at, in pixels, after the viewport-overflow flip — the native
 * port of the web `left` / `top` the `useLayoutEffect` writes onto the menu element.
 */
data class ContextMenuOffset(
    val x: Int,
    val y: Int,
)

/**
 * A pixel size — the measured menu, or the window — used by the placement flip. A small value type so
 * [ContextMenuPlacement.resolvePosition] stays a pure, few-argument function.
 */
data class ContextMenuSize(
    val width: Int,
    val height: Int,
)

/**
 * The active-menu snapshot the host renders — the native port of the web module-store `MenuState`
 * (`{ items, x, y, nonce }`). `null` means closed (the web `state === null`). [nonce] is a monotonic open-counter
 * so re-opening at an identical (items, anchor) still produces a distinct value and re-renders, exactly like the
 * web nonce (e.g. the user right-clicks twice in the same spot). The web `restoreFocusEl` is a DOM concern that
 * Compose's `Popup` focus restoration handles, so it is intentionally not modelled here.
 */
data class ContextMenuState(
    val items: List<ContextMenuItem>,
    val anchor: ContextMenuAnchor,
    val nonce: Long,
) {
    /** True when there are no rows — the open-guard ensures a live store snapshot is never empty. */
    val isEmpty: Boolean
        get() = items.isEmpty()

    /** True when at least one row is interactive (keyboard focus has somewhere to land). */
    val hasEnabledItem: Boolean
        get() = items.any { it.enabled }
}

/**
 * Pure keyboard-roving-focus math — the native port of the web `ContextMenuView` focus helpers
 * (`enabledIndices`, `focusFirstEnabled`, `focusLastEnabled`, `focusNextEnabled`). Disabled rows are skipped and
 * traversal wraps, exactly as the web Arrow Up / Arrow Down / Home / End handlers do. Pure so the contract is
 * asserted off-device and the composable only has to request focus on the index these functions return.
 */
object ContextMenuFocus {
    /** The indices of the interactive (enabled) rows, in display order (web `enabledIndices`). */
    fun enabledIndices(items: List<ContextMenuItem>): List<Int> = items.mapIndexedNotNull { index, item -> index.takeIf { item.enabled } }

    /** The first interactive row, or `null` when every row is disabled (web `focusFirstEnabled`). */
    fun firstEnabledIndex(items: List<ContextMenuItem>): Int? = enabledIndices(items).firstOrNull()

    /** The last interactive row, or `null` when every row is disabled (web `focusLastEnabled`). */
    fun lastEnabledIndex(items: List<ContextMenuItem>): Int? = enabledIndices(items).lastOrNull()

    /**
     * The next interactive row from [current] in the [forward] direction, wrapping at the ends and skipping
     * disabled rows — the native port of the web `focusNextEnabled`. When [current] is not itself an enabled row
     * (e.g. focus is on the menu container), traversal starts at the first enabled row going forward, or the last
     * going backward (the web first-Arrow behaviour). Returns `null` only when there is no enabled row.
     */
    fun nextEnabledIndex(
        items: List<ContextMenuItem>,
        current: Int,
        forward: Boolean,
    ): Int? {
        val enabled = enabledIndices(items)
        if (enabled.isEmpty()) return null
        val cursor = enabled.indexOf(current)
        val resolved =
            if (cursor < 0) {
                if (forward) enabled.first() else enabled.last()
            } else {
                val step = if (forward) 1 else -1
                enabled[(cursor + step + enabled.size) % enabled.size]
            }
        return resolved
    }
}

/**
 * Pure viewport-overflow placement — the native port of the web `ContextMenuView` `useLayoutEffect` flip. Given
 * the captured anchor, the measured menu size, the window size, and the safe [marginPx], it returns the final
 * top-left: when the menu would overflow the right edge it flips to open leftward of the anchor; when it would
 * overflow the bottom edge it flips to open above the anchor; each flipped edge is clamped to at least [marginPx]
 * from the window origin. Pure so the geometry is asserted off-device without a layout pass.
 */
object ContextMenuPlacement {
    /** The safe inset from each window edge (web `VIEWPORT_MARGIN = 8`), expressed in dp at the call site. */
    const val VIEWPORT_MARGIN_DP: Int = 8

    /** Resolves the flipped, clamped top-left for a menu of [menuSize] at the captured [anchor]. */
    fun resolvePosition(
        anchor: ContextMenuAnchor,
        menuSize: ContextMenuSize,
        windowSize: ContextMenuSize,
        marginPx: Int,
    ): ContextMenuOffset {
        val left =
            if (anchor.x + menuSize.width + marginPx > windowSize.width) {
                max(marginPx, anchor.x - menuSize.width)
            } else {
                anchor.x
            }
        val top =
            if (anchor.y + menuSize.height + marginPx > windowSize.height) {
                max(marginPx, anchor.y - menuSize.height)
            } else {
                anchor.y
            }
        return ContextMenuOffset(x = left, y = top)
    }
}

/**
 * The PII-safe diagnostic this surface emits (P1/S11). The `view.opened` event carries only the constant surface
 * [SLUG] — no item labels, no shortcuts, no anchor coordinates — so observability can never leak what was shown
 * or where. Kept free of Compose so it is unit-tested with a recording [Logger].
 */
object ContextMenuDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = CONTEXT_MENU_SLUG

    /** The one-shot event emitted once when the surface (host) opens. */
    const val EVENT_VIEW_OPENED: String = "view.opened"

    /** The structured-field key carrying the surface slug on the diagnostic. */
    const val FIELD_SURFACE: String = "surface"

    /** The structured event emitted when an item handler throws, so a faulty action never breaks the menu. */
    const val EVENT_ITEM_ERROR: String = "contextMenu.itemError"

    /** Emits the one PII-safe `view.opened` diagnostic carrying only the surface [SLUG]. */
    fun recordViewOpened(logger: Logger) {
        logger.info(EVENT_VIEW_OPENED, mapOf(FIELD_SURFACE to SLUG))
    }

    /** Emits the PII-safe item-handler-failure diagnostic — slug only, never the item or its thrown message. */
    fun recordItemError(logger: Logger) {
        logger.warn(EVENT_ITEM_ERROR, mapOf(FIELD_SURFACE to SLUG))
    }
}
